### PR TITLE
Add Logback's throwable-consuming semantics as an option

### LINKED ISSUE #2363: Add Logback's throwable-consuming semantics as an option
## Description
Hello, we used to use logback in our springboot project, recently migrated to log4j2 and we found that a lot of the exception stack is gone. 

For example, in the following example, using logback, you can clearly see the error stack, but log4j2 only has one sentence "java.lang.NullPointerException", I read the source code of log4j2 and feel that this should be a bug. I also test it with version:2.23.1, and the same error.
Additionally, I found that these classes are correct: MessageFormatMessage and StringFormattedMessage.

If you can confirm that this is a bug, I'd be pleasure to submit a PR to fix it.

## Configuration

**Version:** 2.23.1

**Operating system:** windows/Linux

**JDK:** "1.8.0_291"

## Logs

```
==> log4j2 result:

20:32:33.887 [main] [ERROR] com.pandas.springboot.demo.DemoApplicationTests - fail to process null, the error is:java.lang.NullPointerException

==> logback result:
2024-03-11 20:37:35.173 ERROR 16840 --- [           main] c.p.s.demo.DemoApplicationTests          : fail to process null, the error is:{}

java.lang.NullPointerException: null
	at com.pandas.springboot.demo.DemoApplicationTests.log(DemoApplicationTests.java:20) [test-classes/:na]
	at com.pandas.springboot.demo.DemoApplicationTests.testLog(DemoApplicationTests.java:15) [test-classes/:na]
	
```

## Reproduction

```java

@SpringBootTest
public class DemoApplicationTests {

	private static final Logger log = LoggerFactory.getLogger(DemoApplicationTests.class);

	@Test
	public void testLog() {
		log(null);
	}

	private static void log(String param) {
		try {
			int b = param.length();
		} catch (Exception e) {
			log.error("fail to process {}, the error is:{}", param, e);
		}
	}
}

```

### PR DESCRIPTION
**Pull request by ppkarwasz**
The semantics used to determine the `Throwable` attached to a log event changed in Logback 1.1.0 (in the fix to [LOGBACK-873](https://jira.qos.ch/browse/LOGBACK-873)).

When there are not enough arguments to the log call to fill both the throwable field and the placeholders contained in the message pattern, Logback 1.1.0 and later prefer to fill the throwable field and leave some placeholders empty. Log4j Core does the opposite choice: it first fills all the placeholders and only fills the throwable field if there is something left.

This change allows `log4j-slf4j-impl` and `log4j-slf4j2-impl` users to switch between the two behaviors by setting the property `log4j2.messageFactory` to:

```
org.apache.logging.slf4j.message.ThrowableConsumingMessageFactory
```

In this PR two choices have been made:

- the message factory is only included in `log4j-slf4j-impl` and `log4j-slf4j2-impl`, since it should be mostly required by Logback users during a transition period. Note that Logback's and Log4j's semantics to deal with throwables agree whenever enough arguments are provided. In a second step users can modify their code using openrewrite/rewrite-logging-frameworks#141, so that it behaves in the same way in Logback and Log4j Core out-of-the-box. However I can also create a new `log4j-message` artifact for this.
- no **garbage-free** version of this factory is provided. SLF4J creates temporary arrays whenever **more than 2** arguments are provided. I believe that an additional creation of a `ParameterizeMessage` instance does not make a difference.

Closes #2363

### CODE REVIEW

**Review by vy — CHANGES_REQUESTED** [review]
LGTM, thanks for taking care of this @ppkarwasz! :100:
I have two nits:

1. Would you mind explicitly stating in **all** newly added Java source files something like the following, please?
```
This class should be identical to its clone in {@code log4j-slf4j2-impl}.
Any changes made here should be reflected to the other one, and vice versa.
```
2. Would you mind documenting this in the following files, please?
   1. `migrate-from-logback.adoc`
   2. `installation.adoc#impl-logback`

**Comment by ppkarwasz** [comment]
@vy,
>     1. Would you mind explicitly stating in **all** newly added Java source files something like the following, please?
>
> ```
> This class should be identical to its clone in {@code log4j-slf4j2-impl}.
> Any changes made here should be reflected to the other one, and vice versa.
> ```

I don't think this is necessary. Every time one of `log4j-slf4j-impl` or `log4j-slf4j2-impl` is modified, the other one needs to be modified too. 

>     2. Would you mind documenting this in the following files, please?
>        
>        1. `migrate-from-logback.adoc`
>        2. `installation.adoc#impl-logback`

I added some documentation to `migrate-from-logback.adoc` in f7a63206b831043399f04320f533053300114df0.
The `ThrowableConsumingMessageFactory` is only useful for people that migrate from Logback, so I didn't add it to `installation.adoc`.

File: src/changelog/.2.x.x/2363_add_logback_throwable_consuming_semantics.xml [hunk 1]
```diff
@@ -1,8 +1,8 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <entry xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-       xmlns="http://logging.apache.org/log4j/changelog"
-       xsi:schemaLocation="http://logging.apache.org/log4j/changelog https://logging.apache.org/log4j/changelog-0.1.3.xsd"
-       type="added">
+       xmlns="https://logging.apache.org/xml/ns"
+       xsi:schemaLocation="https://logging.apache.org/xml/ns https://logging.apache.org/xml/ns/log4j-changelog-0.xsd"
+       type="changed">
```
Comment (vy) [inline src/changelog/.2.x.x/2363_add_logback_throwable_consuming_semantics.xml:5]: Shouldn't this be `added` instead?

File: src/site/antora/modules/ROOT/examples/migrate-from-logback/MigrateFromLogback.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+import org.apache.logging.log4j.Logger;
+
+class MigrateFromLogback {
+    private static Logger logger = LogManager.getLogger();
+
+    doLogWrong() {
+        try {
+            // do something
+            // tag::wrong[]
+        } catch (Exception e) {
```
Comment (vy) [inline src/site/antora/modules/ROOT/examples/migrate-from-logback/MigrateFromLogback.java:29]: Could you, please, please, please use self-explanatory variable names in the docs? For instance, `error`, `failure`, etc.

File: src/site/antora/modules/ROOT/pages/migrate-from-logback.adoc [hunk 3]
```diff
@@ -106,3 +106,62 @@ To assist with migrating Logback configuration components to Log4j Core, see the
...
+[source]
+----
+The foo process exited with and error.
+java.lang.RuntimeException: Message
+    at example.MigrateFromLogback.doLogWrong(MigrateFromLogback.java:10)
+...
+----
+
+[TIP]
+====
```
Comment (vy) [inline src/site/antora/modules/ROOT/pages/migrate-from-logback.adoc:163]: Block is not closed.

File: src/changelog/.2.x.x/2363_add_logback_throwable_consuming_semantics.xml (hunk 1 above)
Comment (ppkarwasz) [reply src/changelog/.2.x.x/2363_add_logback_throwable_consuming_semantics.xml]: Nice catch! It is a copy/paste error.

Fixed in 42b5ebbb41b4b23f665d395833a10c431a2a602f.

File: src/site/antora/modules/ROOT/examples/migrate-from-logback/MigrateFromLogback.java (hunk 2 above)
Comment (ppkarwasz) [reply src/site/antora/modules/ROOT/examples/migrate-from-logback/MigrateFromLogback.java]: Fixed in 42b5ebbb41b4b23f665d395833a10c431a2a602f.

File: src/site/antora/modules/ROOT/pages/migrate-from-logback.adoc (hunk 3 above)
Comment (ppkarwasz) [reply src/site/antora/modules/ROOT/pages/migrate-from-logback.adoc:163]: Fixed in 42b5ebbb41b4b23f665d395833a10c431a2a602f.

**Review by vy — APPROVED** [review]

**Pull request merged**